core/src/main/java/org/elasticsearch/index/store/Store.java

```diff
@@ -716,6 +719,38 @@ public String toString() {
             return "store(" + in.toString() + ")";
         }
 
+        @Override
+        public void copyFrom(Directory from, String srcFile, String destFile, IOContext context) throws IOException {
+            Directory fromUnwrapped = FilterDirectory.unwrap(from);
+            Directory toUnwrapped = FilterDirectory.unwrap(this);
+            // try to unwrap to FSDirectory - we might be able to just create hard-links of these files and safe copying
+            // the entire file.
+            if (fromUnwrapped instanceof FSDirectory
+                && toUnwrapped instanceof FSDirectory) {
+                final Path fromPath = ((FSDirectory)fromUnwrapped).getDirectory();
+                final Path toPath = ((FSDirectory)toUnwrapped).getDirectory();
+                try {
+                    Files.createLink(toPath.resolve(destFile), fromPath.resolve(srcFile));
+                } catch (FileNotFoundException | NoSuchFileException | FileAlreadyExistsException ex) {
+                    throw ex; // in these cases we bubble up since it's a true error condition.
+                } catch (IOException
+                    | UnsupportedOperationException // if the FS doesn't support hard-links
+                    | SecurityException ex // we might miss a java.nio.file.LinkPermission" "hard"
+                    ) {
+                    // hard-links are not supported or the files are on different filesystems
+                    // we could go deeper and check if their filesstores are the same and opt
+                    // out earlier but for now we just fall back to normal file-copy
+                    try {
+                        super.copyFrom(from, srcFile, destFile, context);
+                    } catch (Exception e) {
+                        e.addSuppressed(ex);
+                        throw ex;
+                    }
+                }
+            } else {
+                super.copyFrom(from, srcFile, destFile, context);
+            }
+        }
     }
```

Comment:
I wonder if Lucene's `Directory.copyFrom` should use hard links (optimization) when it can?  Seems crazy you have to optimize this yourself.